### PR TITLE
EODHP-578 AOI Item Searching

### PR DESCRIPTION
**Pull request by james-hinton**
Extended existing Toolbox functionality to add an item search element to it.

At the moment, this will just take from the users map bounds, in the future, this will be replaced with a polygon draw tool.

### CODE REVIEW

File: src/typings/geojson.d.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
+export type GeoJSONGeometry =
```
Comment (marksmall) [inline src/typings/geojson.d.ts:1]: There is already a package for GeoJSON types at https://www.npmjs.com/package/@types/geojson. It would be better we use those and maybe extend them if we need to add STAC extras, rather than recreating a whole types package.

File: src/utils/genericUtils.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/**
+ *
+ * @param date
+ * @returns string
+ *
+ * This function is used to parse a date in any format into a string.
+ */
+export const parseDate = (date: unknown, includeTime = false): string => {
```
Comment (marksmall) [inline src/utils/genericUtils.tsx:8]: I would say it is better to use a package like [date-fns](https://www.npmjs.com/package/date-fns), rather than build our own which is likely to have missed some gotchas. Dates in JavaScript are notorious.

File: src/utils/genericUtils.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+};
+
+/**
+ *
+ * @param key
+ * @returns string
+ *
+ * This function is used to beautify a key string.
+ */
+export const beautifyKey = (key: string): string => {
```
Comment (marksmall) [inline src/utils/genericUtils.tsx:31]: I would rename this function to something like `titleFromId` as I had to go and find some usage of the function before I really understood what it was for and it is the title that is passed in. The text description could be a bit better to indicate that your taking an ID/Title that has underscores/hyphens and converts that to a space and then uppercasing each word.

File: src/utils/stacUtils.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  return dataPoints;
+};
+
+export const parseFeatureDataPoints = (feature: Feature): DataPoint[] => {
+  // just return the time
+  const dataPoints: DataPoint[] = [];
+  const { properties } = feature;
+  const { datetime } = properties;
```
Comment (marksmall) [inline src/utils/stacUtils.tsx:81]: could be simplified to:
```
const { properties: { datetime } } = feature;
```

File: src/utils/stacUtils.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  return dataPoints;
+};
+
+export const returnFeatureThumbnail = (feature: Feature): string => {
+  return (
+    feature.links.find((link) => link.rel === "thumbnail")?.href ||
+    "https://via.placeholder.com/110"
+  );
```
Comment (marksmall) [inline src/utils/stacUtils.tsx:99]: I'm sure I saw this code in a file further up, or at least very similar, so maybe this function should be extended, if easy to do, to include the `rel` passed in? What is your opinion?

File: tsconfig.json [hunk 6]
```diff
@@ -20,6 +21,9 @@
       ],
       "typings/*": [
         "src/typings/*"
+      ],
+      "utils/*": [
+        "src/utils/*"
```
Comment (marksmall) [inline tsconfig.json:26]: I disagree with adding all these different absolute paths, what are we really saving? All your doing is saving 

`@/utils` to using `utils`, is that a great saving? Keep absolute paths to a minimum, I'd say just the `@` and no more, get rid of `styles`, `typings` and `utils`. This of course should probably be done in a separate PR, I'm not asking for you to change this now.

File: vite.config.ts [hunk 7]
```diff
@@ -10,7 +10,8 @@ export default defineConfig({
     alias: {
       '@': path.resolve(__dirname, './src'),
       'styles': path.resolve(__dirname, './src/styles'),
-      'typings': path.resolve(__dirname, './src/typings')
+      'typings': path.resolve(__dirname, './src/typings'),
+      'utils': path.resolve(__dirname, './src/utils'),
```
Comment (marksmall) [inline vite.config.ts:14]: Same point as above, the more aliases we provide the more config we have and needs parsed, all just to save us typing  `@/`.

**Review by marksmall — COMMENTED** [review]
I think this is way too big a PR, I challenge anyone to review 42 files and fully understand the interactions going on. That being said, at time it is necessary, but I'd like to see PRs much much smaller.

I've made a number of comments, mostly FYI and ideas going forward, along with some coding practices to consider reading up on.

I'm not going to approve this right now to give you a chance to address the comments but I think other than say the **using indices as keys**, I would approve this, with the idea of re-factoring bits of it later.

File: src/components/FilterSidebar/components/TextFilter/types.ts
[the remark this answers is not in the document]
Comment (james-hinton) [reply src/components/FilterSidebar/components/TextFilter/types.ts:1]: I'm fine with either approach, the only worry is about how hard it could be to maintain that file as it grows.

File: src/utils/stacUtils.tsx (hunk 5 above)
Comment (james-hinton) [reply src/utils/stacUtils.tsx:99]: There is another case of this, but it's slightly different.

[Its used on the `onError` callback of an ``](https://github.com/EO-DataHub/eodhp-resource-catalogue-ui/blob/eodhp-578-aoi-item-search/src/pages/MapViewer/components/Toolbox/components/ToolboxRow/index.tsx#L18)

As we don't host the images, sometimes a thumbnail link provided may be behind a paywall or unavailable in some way. Should we still try to figure out a way to merge the functions?
